src/blockchain_db/lmdb/db_lmdb.cpp

```diff
@@ -460,49 +471,279 @@ void mdb_txn_safe::allow_new_txns()
+mdb_txn_cursors& mdb_writer_data::cursors()
+{
+  return *m_cursors;
+}
 
-  mdb_env_info(env, &mei);
-  uint64_t new_mapsize = mei.me_mapsize;
+MDB_cursor* mdb_writer_data::cursor(source database)
+{
+  if (current_thread() != boost::this_thread::get_id())
+    throw1(DB_ERROR{"transaction owned by other threayyd"});
 
-  MGINFO("LMDB Mapsize increased." << "  Old: " << old / (1024 * 1024) << "MiB" << ", New: " << new_mapsize / (1024 * 1024) << "MiB");
+  auto& cursor = m_cursors->get(database);
 
-  mdb_txn_safe::allow_new_txns();
+  if (cursor != nullptr)
+    return cursor;
+
+  auto res = mdb_cursor_open(m_txn, m_databases.get(database), &cursor);
+
+  if (res == 0)
+    return cursor;
+
+  auto error_message =
+    "Failed to open a cursor for " +
+    mdb_databases::name(database) +
+    ": " +
+    mdb_strerror(res);
+
+  throw0(DB_ERROR(error_message.c_str()));
 }
 
-inline int lmdb_txn_begin(MDB_env *env, MDB_txn *parent, unsigned int flags, MDB_txn **txn)
+MDB_txn* mdb_writer_data::try_acquire() noexcept
 {
-  int res = mdb_txn_begin(env, parent, flags, txn);
-  if (res == MDB_MAP_RESIZED) {
-    lmdb_resized(env);
-    res = mdb_txn_begin(env, parent, flags, txn);
+  if (current_thread() != boost::this_thread::get_id())
+    return nullptr;
+
+  ++m_writer_depth;
+  return m_txn;
+}
+
+MDB_txn* mdb_writer_data::acquire()
+{
+  LOG_PRINT_L3("mdb_writer_data: acquire()");
+
+  if (auto txn = try_acquire())
+    return txn;
+
+  while (mdb_txn_safe::creation_gate.test_and_set());
+  mdb_txn_safe::num_active_txns++;
+  mdb_txn_safe::creation_gate.clear();
+
+  auto result = lmdb_txn_begin(m_env, nullptr, 0, &m_txn);
+
+  if (result != 0)
+    throw0(DB_ERROR{lmdb_error("Failed to create a transaction for the db: ", result).c_str()});
+
+  m_cursors.emplace();
+  m_writer_thread.store(boost::this_thread::get_id(), boost::memory_order_relaxed);
+  m_writer_depth = 1;
+
+  return m_txn;
+}
+
+void mdb_writer_data::release(bool error)
+{
+  if (current_thread() != boost::this_thread::get_id())
+    throw1(DB_ERROR{"transaction owned by other threayyd"});
+
+  if (m_writer_depth == 0)
+    throw1(DB_ERROR{"transaction already released"});
+
+  if (error)
+    m_error = error;
+
+  if (--m_writer_depth != 0 || m_txn == nullptr)
+    return;
+
+  auto txn = m_txn;
+  m_txn = nullptr;
+  m_cursors.reset();
+  m_writer_thread.store({}, boost::memory_order_relaxed);
+
+  if (m_error)
+  {
+    mdb_txn_abort(txn);
+    mdb_txn_safe::num_active_txns--;
+  }
+  else
+  {
+    auto result = mdb_txn_commit(txn);
+    mdb_txn_safe::num_active_txns--;
+
+    if (result)
+      throw0(DB_ERROR{mdb_strerror(result)});
   }
-  return res;
 }
 
-inline int lmdb_txn_renew(MDB_txn *txn)
+mdb_write_handle::~mdb_write_handle()
 {
-  int res = mdb_txn_renew(txn);
-  if (res == MDB_MAP_RESIZED) {
-    lmdb_resized(mdb_txn_env(txn));
-    res = mdb_txn_renew(txn);
+  if (m_data == nullptr || m_txn == nullptr)
```

Comment:
I notice a `std::current_exception()` which clearly indicates this is designed to grab errors as the stack is being unwound. But this function also potentially throws on the last two lines, so this will need to be an empty `try{} catch {}` to prevent an immediate process abortion.

Reply:
Interesting point. Isn't the fact that logging can throw a potentially really big problem in our code-base? We do seem to log from destructors in many other places which basically means these can potentially throw as well, which is rather iffy.

I fixed it here, also because the `release()` can throw if committing the transaction fails.

Reply:
Moo has been adding empty catch blocks in some places. It kind of stinks because you can't really even log, unless its a double catch or a `noexcept` function. The more likely case is a throw from `release(...)` though.

Reply:
I never answered your first question. Yes, a `noexcept` log function could be useful because its more intuitive even if logging in a destructor fails. Big uphill battle on that  though.

Reply:
I take that back, it should be a matter of wrapping the call in a catch block. The issue is that it would return an error code that isn't checked anywhere. And there are cases where nothing could be done on error anyway (which is basically what this is about).